src/main/java/org/sciborgs1155/robot/commands/NoteVisualizer.java

```diff
@@ -0,0 +1,190 @@
+import edu.wpi.first.networktables.StructArrayTopic;
+import edu.wpi.first.networktables.StructPublisher;
+import edu.wpi.first.networktables.StructTopic;
+import edu.wpi.first.wpilibj2.command.Command;
+import edu.wpi.first.wpilibj2.command.Commands;
+import edu.wpi.first.wpilibj2.command.ScheduleCommand;
+import java.util.ArrayList;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Set;
+import java.util.function.DoubleSupplier;
+import java.util.function.Supplier;
+import monologue.Logged;
+import org.sciborgs1155.robot.shooter.ShooterConstants;
+
+public class NoteVisualizer implements Logged {
+  // notes
+  private static LinkedList<Pose3d> notes =
+      new LinkedList<>(
+          List.of(
+              BLUE_LEFT_NOTE,
+              BLUE_MID_NOTE,
+              BLUE_RIGHT_NOTE,
+              CENTER_NOTE_ONE,
+              CENTER_NOTE_TWO,
+              CENTER_NOTE_THREE,
+              CENTER_NOTE_FOUR,
+              CENTER_NOTE_FIVE,
+              RED_LEFT_NOTE,
+              RED_MID_NOTE,
+              RED_RIGHT_NOTE));
+  private static boolean carryingNote = false;
+  private static ArrayList<Pose3d> pathPosition = new ArrayList<>();
+
+  // suppliers
+  private static Supplier<Pose2d> pose = Pose2d::new;
+  private static Supplier<Rotation3d> angle = Rotation3d::new;
+  private static DoubleSupplier velocity = () -> 1;
+
+  private static double zVelocity;
+  private static int i = 0;
+
+  private static Pose3d currentNotePose = new Pose3d();
+  private static Pose3d lastNotePose = new Pose3d();
+
+  // publishers
+  private static StructPublisher<Pose3d> shotNotePub;
+  private static StructArrayPublisher<Pose3d> pathPub;
+  private static StructArrayPublisher<Pose3d> notesPub;
+  private static BooleanPublisher hasNotePub;
+
+  public static void setSuppliers(
+      Supplier<Pose2d> robotPose, Supplier<Rotation3d> pivotAngle, DoubleSupplier shotVelocity) {
+    pose = robotPose;
+    angle = pivotAngle;
+    velocity = shotVelocity;
+  }
+
+  /** Set up NT publisher. Call only once before beginning to log notes. */
+  public static void startPublishing() {
+    NetworkTableInstance inst = NetworkTableInstance.getDefault();
+
+    StructArrayTopic<Pose3d> notesTopic = inst.getStructArrayTopic("notes", Pose3d.struct);
+    StructArrayTopic<Pose3d> notePathTopic = inst.getStructArrayTopic("note path", Pose3d.struct);
+    BooleanTopic hasNoteTopic = inst.getBooleanTopic("holding note");
+    StructTopic<Pose3d> shotNoteTopic = inst.getStructTopic("shot note", Pose3d.struct);
+
+    notesPub = notesTopic.publish();
+    pathPub = notePathTopic.publish();
+    hasNotePub = hasNoteTopic.publish();
+    shotNotePub = shotNoteTopic.publish();
+
+    notesPub.set(notes.toArray(new Pose3d[0]));
+  }
+
+  public static void log() {
+    shotNotePub.set(currentNotePose);
+    hasNotePub.set(carryingNote);
+  }
+
+  public static Command intake() {
+    if (carryingNote) {
```

Comment:
this is executed at initialization, not runtime, since it isn't inside the command